Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean

```diff
@@ -0,0 +1,967 @@
+    (hPz : P z ≠ 0) (hQz : Q z ≠ 0) (hx : P x * Q z = P z * Q x) (hy : P y * Q z ≠ P z * W.negY Q) :
+    W.addZ_of_Yne P ≠ 0 :=
+  mul_ne_zero (pow_ne_zero 3 hPz) <| pow_ne_zero 3 <| sub_ne_zero_of_ne <|
+    Yne_of_Yne hP hQ hPz hQz hx hy
+
+lemma addY_div_addZ_of_Xne {P Q : Fin 3 → F} (hPz : P z ≠ 0) (hQz : Q z ≠ 0)
+    (hx : P x * Q z ≠ P z * Q x) : W.addY_of_Xne P Q / addZ_of_Xne P Q =
+      W.toAffine.addY (P x / P z) (Q x / Q z) (P y / P z)
+        (W.toAffine.slope (P x / P z) (Q x / Q z) (P y / P z) (Q y / Q z)) := by
+  rw [Affine.addY, ← addX_div_addZ_of_Xne hPz hQz hx, ← addY'_div_addZ_of_Xne hPz hQz hx]
+  exact negY_divZ <| addZ_ne_zero_of_Xne hPz hQz hx
+
+lemma addY_div_addZ_of_Yne {P Q : Fin 3 → F} (hP : W.nonsingular P) (hQ : W.nonsingular Q)
+    (hPz : P z ≠ 0) (hQz : Q z ≠ 0) (hx : P x * Q z = P z * Q x) (hy : P y * Q z ≠ P z * W.negY Q) :
+    W.addY_of_Yne P / W.addZ_of_Yne P = W.toAffine.addY (P x / P z) (Q x / Q z) (P y / P z)
+      (W.toAffine.slope (P x / P z) (Q x / Q z) (P y / P z) (Q y / Q z)) := by
+  rw [Affine.addY, ← addX_div_addZ_of_Yne hP hQ hPz hQz hx hy,
+    ← addY'_div_addZ_of_Yne hP hQ hPz hQz hx hy]
+  exact negY_divZ <| addZ_ne_zero_of_Yne hP hQ hPz hQz hx hy
+
+end Polynomial
+
+section Representative
+
+/-! ### Group operations on point representatives -/
+
+/-- The negation of a point representative. -/
+@[pp_dot]
+def neg (P : Fin 3 → R) : Fin 3 → R :=
+  ![P x, W.negY P, P z]
+
+@[simp]
+lemma neg_zero : W.neg ![0, 1, 0] = ![0, -1, 0] := by
+  erw [neg, negY, mul_zero, mul_zero, sub_zero, sub_zero]
+  rfl
+
+@[simp]
+lemma neg_some (X Y : R) : W.neg ![X, Y, 1] = ![X, -Y - W.a₁ * X - W.a₃, 1] := by
+  erw [neg, negY, mul_one]
+  rfl
+
+lemma neg_smul_equiv (P : Fin 3 → R) (u : Rˣ) : W.neg (u • P) ≈ W.neg P :=
+  ⟨u, by simp_rw [neg, negY_smul, smul_fin3]; rfl⟩
+
+lemma neg_equiv {P Q : Fin 3 → R} (h : P ≈ Q) : W.neg P ≈ W.neg Q := by
+  rcases h with ⟨u, rfl⟩
+  exact W.neg_smul_equiv Q u
+
+/-- The negation of a point class. If `P` is a point representative,
+then `W.neg_map ⟦P⟧` is definitionally equivalent to `W.neg P`. -/
+@[pp_dot]
+def neg_map (P : PointClass R) : PointClass R :=
+  P.map W.neg fun _ _ => W.neg_equiv
+
+lemma neg_map_eq {P : Fin 3 → R} : W.neg_map ⟦P⟧ = ⟦W.neg P⟧ :=
+  rfl
+
+@[simp]
+lemma neg_map_zero : W.neg_map ⟦![0, 1, 0]⟧ = ⟦![0, 1, 0]⟧ := by
+  simpa only [neg_map_eq, neg_zero, Quotient.eq] using ⟨-1, by simp⟩
+
+@[simp]
+lemma neg_map_some (X Y : R) : W.neg_map ⟦![X, Y, 1]⟧ = ⟦![X, -Y - W.a₁ * X - W.a₃, 1]⟧ := by
+  rw [neg_map_eq, neg_some]
+
+open scoped Classical
+
+/-- The addition of two point representatives. -/
+@[pp_dot]
+noncomputable def add (P Q : Fin 3 → R) : Fin 3 → R :=
+  if P z = 0 then Q else if Q z = 0 then P else if P x * Q z = P z * Q x then
+    if P y * Q z = P z * W.negY Q then ![0, 1, 0] else
+      ![W.addX_of_Yne P, W.addY_of_Yne P, W.addZ_of_Yne P]
+  else ![W.addX_of_Xne P Q, W.addY_of_Xne P Q, addZ_of_Xne P Q]
```

Comment:
I think these could be called `![W.dblX P, W.dblY P, W.dblZ P]` and `![W.addX P, W.addY P, addZ P]` just like the Jacobian case. In fact, I've found that you can remove a `P z ^ 2` factor from the projective doubling formula to make it valid in all cases (O+O or P+P for P 2-torsion); the resulting formula is homogeneous of degree 4 rather than 6. From the projective addition formula, you can also remove a factor of `P z * Q z`, and the resulting formula is homogeneous of degree 6 rather than 8, and applies in the case P=-Q≠Q, but not in the cases P=O or Q=O (gives (0,0,0)). I don't think this simplifies the proof that the reduction map is a homomorphism though.

I think this is the best you can do in projective coordinates, which is not as nice as the Jacobian formulas, where both doubling and addition are weighted homogeneous of degree (8,12,4) and applies more universally without needing to single out the cases P+O and O+P.

The way to remove the factors is as follows: for `Xne`/`add`, subtract $z_1^3 F(x_2,y_2,z_2)+z_2^3 F(x_1,y_1,z_1)$ (where $F$ is W.polynomial) from `addX'_of_Xne`, then it becomes divisible by $z_1 z_2$. Plug this into `addY'_of_Xne` and add $-y_1  z_1^3 z_2 F(x_2, y_2, z_2) + y_2 z_1 z_2^3 F(x_1, y_1, z_1)$, then it also becomes divisible by $z_1 z_2$ (in fact even by its square). For the `Yne`/`dbl`, add $9x F(x,y,z)$ to `addX'_of_Yne` then it becomes divisible by $z$; the other factor of $z$ is added when passing to `addX_of_Yne`. Plug this new version of `addX'_of_Yne` + $(a_1^2 + 3 a_2)F(x,y,z)$ into `addY'_of_Yne` and subtract $z(3 a_1^2  x^2 + 9 a_1  x  y + 9 y^2) P[x, y, z]$, then `addY'_of_Yne` becomes divisible by $z^2$.

Reply:
Thanks! It turns out that you can still do a bit better for projective addition and make everything homogeneous of degree 4 (with slightly less appealing expressions). This makes the formula uniform when precisely one of `P` or `Q` is `O` (not both, but we don't need this since it's part of `dblXYZ`, which is the same for `Jacobian`). I've done this in #9417

Reply:
Hmm thanks! I can reproduce ... so this should give rise to multiplication-by-n formula in projective coordinates, and I wonder how it compares with $(\phi\psi:\omega:\psi^3)$ ...